Biome plugin not working with WSL

I tried to use the plugin with a project in Windows Subsystem for Linux. I couldn't get it to work and tried to gather some relevant log entries from the IDEA logs.

First I tried with `Automatic Biome configuration`. The plugin seemingly loads fine, no errors. However the formatter hangs on save and I got this in the logs:

```
2024-04-10 19:17:24,713 [ 777798] SEVERE - #com.github.biomejs.intellijbiome.actions.BiomeCheckRunner - Failed to run Biome check with (format, safefixes) for file App.jsx - Timeout exceeded
java.lang.Throwable: Failed to run Biome check with (format, safefixes) for file App.jsx - Timeout exceeded
        at com.intellij.openapi.diagnostic.Logger.error(Logger.java:376)
        at com.github.biomejs.intellijbiome.actions.BiomeCheckRunner.applyChanges(BiomeCheckRunner.kt:91)
        at com.github.biomejs.intellijbiome.actions.BiomeCheckRunner.access$applyChanges(BiomeCheckRunner.kt:20)
        at com.github.biomejs.intellijbiome.actions.BiomeCheckRunner$formatDocuments$1$1.run(BiomeCheckRunner.kt:58)
        at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:477)
        at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:133)
--
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
        at java.base/java.lang.Thread.run(Thread.java:840)
```

Then I tried with `Manual Biome configuration` pointing executable and biome.json to files inside WSL. The plugin immediately indicates the language server is terminated (screenshot below). In addition when saving any file the log contains entries about invalid path (see below).

![image](https://github.com/biomejs/biome-intellij/assets/8472329/e1f2c097-01c5-4882-a95d-2ede55acf7fe)

```
2024-04-10 19:25:52,261 [1285346] SEVERE - #c.i.o.p.Task - Cannot run program "\\wsl.localhost\Ubuntu-22.04\home\niklas\code\my-project\node_modules\.bin\biome" (in directory "\\wsl.localhost\Ubuntu-22.04\home\niklas\code\my-project"): CreateProcess error=3, The system cannot find the path specified
com.intellij.execution.process.ProcessNotCreatedException: Cannot run program "\\wsl.localhost\Ubuntu-22.04\home\niklas\code\my-project\node_modules\.bin\biome" (in directory "\\wsl.localhost\Ubuntu-22.04\home\niklas\code\my-project"): CreateProcess error=3, The system cannot find the path specified
        at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:366)
        at com.intellij.execution.process.OSProcessHandler.startProcess(OSProcessHandler.java:86)
        at com.intellij.execution.process.OSProcessHandler.<init>(OSProcessHandler.java:46)
        at com.intellij.execution.process.CapturingProcessHandler.<init>(CapturingProcessHandler.java:20)
        at com.github.biomejs.intellijbiome.extensions.CommandLineExtKt.runProcessFuture(CommandLineExt.kt:26)
        at com.github.biomejs.intellijbiome.BiomeStdinRunner.check(BiomeStdinRunner.kt:34)
        at com.github.biomejs.intellijbiome.actions.BiomeCheckRunner$formatDocuments$1$1.run(BiomeCheckRunner.kt:55)
        at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:477)
        at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:133)
--
        at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.io.IOException: Cannot run program "\\wsl.localhost\Ubuntu-22.04\home\niklas\code\my-project\node_modules\.bin\biome" (in directory "\\wsl.localhost\Ubuntu-22.04\home\niklas\code\my-project"): CreateProcess error=3, The system cannot find the path specified
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1143)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
--
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1110)
        ... 34 more
```

Probably requires some additional handling, but I guess it should be doable since IDEA works nicely with WSL.

On a side note: would be worth it to add some error handling too, because especially with the automatic configuration there's no errors so the plugin just shows as not working or buggy for the user.

---

Environment information:

IntelliJ IDEA 2024.1 (Ultimate Edition)
Biome plugin v1.0.0

--- a/gradle.properties
+++ b/gradle.properties
@@ -11,7 +11,7 @@ pluginUntilBuild = 243.*
 
 # IntelliJ Platform Properties -> https://plugins.jetbrains.com/docs/intellij/tools-gradle-intellij-plugin.html#configuration-intellij-extension
 platformType = WS
-platformVersion = 2024.3
+platformVersion = 2024.3.3
 
 # Plugin Dependencies -> https://plugins.jetbrains.com/docs/intellij/plugin-dependencies.html
 # Example: platformPlugins = com.jetbrains.php:203.4449.22, org.intellij.scala:2023.3.27@EAP

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/BiomePackage.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/BiomePackage.kt
@@ -93,7 +93,11 @@ class BiomePackage(private val project: Project) {
         }
 
         val processHandler =
-            BiomeTargetRunBuilder(project).getBuilder(binaryPath).addParameters(listOf("--version")).build()
+            BiomeTargetRunBuilder(project)
+                .getBuilder(binaryPath)
+                .addParameters(listOf(ProcessCommandParameter.Value("--version")))
+                .build()
+
         return runCatching {
             val result = runProcessFuture(processHandler).await()
             val processOutput = result.processOutput

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/GeneralProcessCommandBuilder.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/GeneralProcessCommandBuilder.kt
@@ -29,8 +29,9 @@ class GeneralProcessCommandBuilder : ProcessCommandBuilder {
         return this
     }
 
-    override fun addParameters(params: List<String>): ProcessCommandBuilder {
-        parameters.addAll(params)
+    override fun addParameters(params: List<ProcessCommandParameter>): ProcessCommandBuilder {
+        // TODO: custom executable is not supported in WSL yet
+        parameters.addAll(params.map { it.toString() })
         return this
     }
 

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/NodeProcessCommandBuilder.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/NodeProcessCommandBuilder.kt
@@ -33,8 +33,13 @@ class NodeProcessCommandBuilder(
         return this
     }
 
-    override fun addParameters(params: List<String>): ProcessCommandBuilder {
-        parameters.addAll(params)
+    override fun addParameters(params: List<ProcessCommandParameter>): ProcessCommandBuilder {
+        parameters.addAll(params.map {
+            when (it) {
+                is ProcessCommandParameter.Value -> it.value
+                is ProcessCommandParameter.FilePath -> target.convertLocalPathToTargetPath(it.path)
+            }
+        })
         return this
     }
 

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/ProcessCommandBuilder.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/ProcessCommandBuilder.kt
@@ -7,7 +7,7 @@ import java.nio.charset.Charset
 interface ProcessCommandBuilder {
     fun setWorkingDirectory(path: String?): ProcessCommandBuilder
     fun setInputFile(file: VirtualFile?): ProcessCommandBuilder
-    fun addParameters(params: List<String>): ProcessCommandBuilder
+    fun addParameters(params: List<ProcessCommandParameter>): ProcessCommandBuilder
     fun setExecutable(executable: String): ProcessCommandBuilder
     fun setCharset(charset: Charset): ProcessCommandBuilder
     fun build(): OSProcessHandler

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/ProcessCommandParameter.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/ProcessCommandParameter.kt
@@ -1,0 +1,22 @@
+package com.github.biomejs.intellijbiome
+
+/**
+ * A command argument for creating a process.
+ */
+sealed interface ProcessCommandParameter {
+    /**
+     * A raw string value.
+     * No conversion will be applied on any environment.
+     */
+    class Value(val value: String) : ProcessCommandParameter {
+        override fun toString() = value
+    }
+
+    /**
+     * A file path value which points any file or directory on the execution environment.
+     * For example, the path will be converted automatically when running Node.js on WSL 2.
+     */
+    class FilePath(val path: String) : ProcessCommandParameter {
+        override fun toString() = path
+    }
+}

--- a/src/main/kotlin/com/github/biomejs/intellijbiome/lsp/BiomeLspServerSupportProvider.kt
+++ b/src/main/kotlin/com/github/biomejs/intellijbiome/lsp/BiomeLspServerSupportProvider.kt
@@ -3,6 +3,7 @@ package com.github.biomejs.intellijbiome.lsp
 import com.github.biomejs.intellijbiome.BiomeIcons
 import com.github.biomejs.intellijbiome.BiomePackage
 import com.github.biomejs.intellijbiome.BiomeTargetRunBuilder
+import com.github.biomejs.intellijbiome.ProcessCommandParameter
 import com.github.biomejs.intellijbiome.services.BiomeServerService
 import com.github.biomejs.intellijbiome.settings.BiomeConfigurable
 import com.github.biomejs.intellijbiome.settings.BiomeSettings
@@ -57,10 +58,10 @@ import com.intellij.util.SmartList
     }
 
     override fun startServerProcess(): OSProcessHandler {
-        val params = SmartList("lsp-proxy")
+        val params = SmartList<ProcessCommandParameter>(ProcessCommandParameter.Value("lsp-proxy"))
         if (!configPath.isNullOrEmpty()) {
-            params.add("--config-path")
-            params.add(configPath)
+            params.add(ProcessCommandParameter.Value("--config-path"))
+            params.add(ProcessCommandParameter.FilePath(configPath))
         }
 
         return targetRunBuilder.getBuilder(executable).apply {

